extensions/wikia/AdEngine/js/provider/gpt/GoogleTag.js

```diff
@@ -0,0 +1,149 @@
+/*global define,setTimeout*/
+/*jshint maxlen:125, camelcase:false, maxdepth:7*/
+define('ext.wikia.adEngine.provider.gpt.googleTag', [
+	'wikia.document',
+	'wikia.log',
+	'wikia.window'
+], function (doc, log, window) {
+
+	var logGroup = 'ext.wikia.adEngine.provider.gpt.googleTag',
+		initialized = false,
+		registeredCallbacks = {},
+		slots = {},
+		slotQueue = [],
+		pageLevelParams,
+		googleTag,
+		pubAds;
+
+	function init() {
+		log('init', 'debug', logGroup);
+
+		var gads = doc.createElement('script'),
+			node = doc.getElementsByTagName('script')[0];
+
+		window.googletag = window.googletag || {};
+		window.googletag.cmd = window.googletag.cmd || [];
+
+		gads.async = true;
+		gads.type = 'text/javascript';
+		gads.src = '//www.googletagservices.com/tag/js/gpt.js';
+
+		log('Appending GPT script to head', 'debug', logGroup);
+
+		node.parentNode.insertBefore(gads, node);
+		googleTag = window.googletag;
+
+		// Enable services
+		log(['init', 'googletag.cmd.push'], 'info', logGroup);
+		googleTag.cmd.push(function () {
+			pubAds = googleTag.pubads();
+
+			pubAds.collapseEmptyDivs();
+			pubAds.enableSingleRequest();
+			pubAds.disableInitialLoad(); // manually request ads using refresh
+			pubAds.addEventListener('slotRenderEnded', dispatchEvent);
+
+			googleTag.enableServices();
+
+			log(['init', 'googletag.cmd.push', 'done'], 'debug', logGroup);
+		});
+		initialized = true;
+	}
+
+	function isInitialized() {
+		return initialized;
+	}
+
+	function setPageLevelParams(params) {
+		var name,
+			value;
+
+		googleTag.cmd.push(function () {
+			pageLevelParams = params;
+			for (name in pageLevelParams) {
+				if (pageLevelParams.hasOwnProperty(name)) {
+					value = pageLevelParams[name];
+					if (value) {
+						log(['setPageLevelParams', 'pubAds.setTargeting', name, value], 'debug', logGroup);
+						pubAds.setTargeting(name, value);
+					}
+				}
+			}
+		});
+	}
+
+	function push(callback) {
+		googleTag.cmd.push(callback);
+	}
+
+	function flush() {
+		if (!isInitialized()) {
+			log(['flush', 'done', 'No slots to flush'], 'info', logGroup);
+			return;
+		}
+
+		googleTag.cmd.push(function () {
+			log(['flush', 'start'], 'info', logGroup);
+
+			log(['flush', 'refresh', slotQueue], 'debug', logGroup);
+			if (slotQueue.length) {
+				pubAds.refresh(slotQueue);
+				slotQueue = [];
+			}
+
+			log(['flush', 'done'], 'info', logGroup);
+		});
+	}
+
+	function addSlot(element) {
```

Comment:
Check example here: 
https://github.com/Wikia/app/blob/dev/extensions/wikia/AdEngine/js/slot/inContent.js#L14

On 06/19/2015 02:56 PM, Andrzej wrote:
> In extensions/wikia/AdEngine/js/provider/gpt/GoogleTag.js
> <https://github.com/Wikia/app/pull/7565#discussion_r32825770>:
>
>> +		}
>> +
>> +		googleTag.cmd.push(function () {
>> +			log(['flush', 'start'], 'info', logGroup);
>> +
>> +			log(['flush', 'refresh', slotQueue], 'debug', logGroup);
>> +			if (slotQueue.length) {
>> +				pubAds.refresh(slotQueue);
>> +				slotQueue = [];
>> +			}
>> +
>> +			log(['flush', 'done'], 'info', logGroup);
>> +		});
>> +	}
>> +
>> +	function addSlot(element) {
>
> It'd be useful to put here information (JsDoc?) about what's the
> element's interface since we definitely want it to have|getId()|,
> |setPageLevelParams()|, |getSlotPath()| and |getSizes()| methods.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Wikia/app/pull/7565/files#r32825770>.
>